### PR TITLE
Add spacelift_policies data source

### PR DESCRIPTION
**Pull request by alexjurkiewicz**
This can be used to find all policies of a certain type (PLAN, ACCESS, etc) with specific tags.

One example use-case is in attaching all policies with a specific tag to new stacks:

```terraform
resource "spacelift_stack" "this" {
 # ...
}

data "spacelift_policies" "plan_autoattach" {
 type = "PLAN"
 tags = ["standard_policies"]
}

resource "spacelift_policy_attachment" "this" {
 for_each  = data.spcelift_policies.plan_autoattach
 policy_id = each.value.id
 stack_id  = spacelift_stack.this.id
}
```

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (non-breaking change that adds documentation)

### Development

- [x] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review

- [x] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [x] Pull Request is no longer marked as "draft"
- [ ] Reviewers have been assigned
- [ ] Changes have been reviewed by at least one other engineer
- [ ] The target branch is `future` unless the change is going directly into production

### CODE REVIEW

File: spacelift/data_policies.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+			"type": {
+				Type:        schema.TypeString,
+				Description: "type of the policy",
+				Required:    true,
+			},
+			"labels": {
+				Type:        schema.TypeSet,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "required labels to match",
+				Required:    true,
```
Comment (marcinwyszynski) [inline spacelift/data_policies.go:31]: I would not mark it as required. If someone doesn't want to match by label (which is a valid case IMO), we can allow them to not specify any labels.

File: spacelift/data_policies.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+			"`spacelift_policies` can find all policies that have certain labels," +
+			"and return their IDs. ",
+
+		ReadContext: dataPoliciesRead,
+
+		Schema: map[string]*schema.Schema{
+			"type": {
+				Type:        schema.TypeString,
+				Description: "type of the policy",
+				Required:    true,
```
Comment (marcinwyszynski) [inline spacelift/data_policies.go:25]: I would not mark it as required. Someone may want to match using labels instead.

File: spacelift/data_policies.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	requestedType := typeRaw.(string)
+	labelsRaw, labelsSpecified := d.GetOk("labels")
+	requestedLabels := labelsRaw.(*schema.Set).List()
+	ids := schema.NewSet(schema.HashString, []interface{}{})
+	for _, policy := range query.Policies.Policies {
+		if typeSpecified && policy.Type != requestedType {
+			continue
+		}
+		if labelsSpecified {
+			for _, required := range requestedLabels {
```
Comment (alexjurkiewicz) [inline spacelift/data_policies.go:59]: I'm not sure about this method to check all requested labels are present. Is there a better way?

File: spacelift/data_policies.go (hunk 1 above)
Comment (alexjurkiewicz) [reply spacelift/data_policies.go]: done, labels are optional

File: spacelift/data_policies.go (hunk 2 above)
Comment (alexjurkiewicz) [reply spacelift/data_policies.go]: done, type is optional

**Comment by alexjurkiewicz** [comment]
Three questions:

1. What's the process to generate docs?
2. What's the process to run only my new acceptance tests locally?
3. What's the next step?

File: spacelift/internal/testhelpers/helpers.go [hunk 4]
```diff
@@ -207,3 +207,61 @@ func SetEquals(name string, values ...string) AttributeCheck {
 		return nil
 	}
 }
+
+// SetContains checks the set contains all specified values
+func SetContains(name string, values ...string) AttributeCheck {
```
Comment (alexjurkiewicz) [inline spacelift/internal/testhelpers/helpers.go:212]: these three set functions are almost fully duplicated, is that a problem?

File: spacelift/data_policies.go (hunk 3 above)
Comment (marcinwyszynski) [reply spacelift/data_policies.go:59]: Looks good.

File: spacelift/internal/testhelpers/helpers.go (hunk 4 above)
Comment (marcinwyszynski) [reply spacelift/internal/testhelpers/helpers.go:212]: Not a problem, no.

**Review by marcinwyszynski — COMMENTED** [review]
As requested previously, please have the enumeration return a richer collection, not just a list of IDs. Otherwise LGTM.

**Comment by marcinwyszynski** [comment]
> What's the process to generate docs?

We use [this tool](https://github.com/hashicorp/terraform-plugin-docs).

> What's the process to run only my new acceptance tests locally?

You can populate [these 3 variables](https://github.com/spacelift-io/terraform-provider-spacelift/blob/master/.github/workflows/test.yml#L30-L32) to point at your account and just run this single unit test.

> What's the next step?

Once the suggestions are implemented, we will merge this PR and create a new provider release.

**Comment by alexjurkiewicz** [comment]
> As requested previously, please have the enumeration return a richer collection, not just a list of IDs. Otherwise LGTM.

I was modelling this data source on `aws_subnet_ids` ([link](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/subnet_ids)) which only returns the list of IDs. But do you suggest I return a map of policies with more metadata? Like returning `map(object({name=string,type=string,labels=list(string)}))` (and the policy ID is the map key).

**Comment by alexjurkiewicz** [comment]
I ended up writing my own smoke test to be sure this data source works as expected. Here's an example Terraform configuration that tests all four possible states:

1. Load all policies
2. Load policies of a given type
3. Load policies with a given tag
4. Load policies of a given type with a given tag

It worked 😊 

<details>
  <summary>main.tf</summary>

```terraform
terraform {
  required_providers {
    spacelift = {
      source = "spacelift-io/spacelift"
    }
  }
}
provider "spacelift" {}

resource "spacelift_policy" "test" {
  count  = 5
  name   = "test ${count.index}"
  labels = ["spacelift_policies_test"]
  type   = "ACCESS"
  body   = <<EOF
package spacelift
read { true}
EOF
}
resource "spacelift_policy" "test2" {
  name   = "test 5"
  labels = ["spacelift_policies_test"]
  type   = "GIT_PUSH"
  body   = <<EOF
package spacelift
track { true }
EOF
}

data "spacelift_policies" "all" {}
data "spacelift_policies" "type" {
  type = "LOGIN"
}
data "spacelift_policies" "label" {
  labels     = ["spacelift_policies_test"]
  depends_on = [spacelift_policy.test, spacelift_policy.test2]
}
data "spacelift_policies" "labelled_access" {
  type       = "ACCESS"
  labels     = ["spacelift_policies_test"]
  depends_on = [spacelift_policy.test, spacelift_policy.test2]
}

output "number_of_total_policies" {
  value = length(data.spacelift_policies.all.policies)
}
output "login_policy" {
  value = data.spacelift_policies.type.policies[0]
}
output "number_of_labelled_policies" {
  value = length(data.spacelift_policies.label.policies)
}
output "number_of_labelled_access_policies" {
  value = length(data.spacelift_policies.labelled_access.policies)
}
```
</details>

<details>
  <summary>Apply output</summary>

```
Apply complete! Resources: 6 added, 0 changed, 0 destroyed.

Outputs:

login_policy = {
  "id" = "login-policy"
  "labels" = toset([])
  "name" = "login-policy"
  "type" = "LOGIN"
}
number_of_labelled_access_policies = 5
number_of_labelled_policies = 6
number_of_total_policies = 1028
```

</details>

**Comment by alexjurkiewicz** [comment]
I tried generating docs with v0.8.1 of `terraform-plugin-docs`, but it tried to rewrite every docs page, it looks like there have been substantial changes to the output format. So I haven't committed any docs page.

**Comment by adamconnelly** [comment]
> I tried generating docs with v0.8.1 of `terraform-plugin-docs`, but it tried to rewrite every docs page, it looks like there have been substantial changes to the output format. So I haven't committed any docs page.

@alexjurkiewicz I've just merged a change to update to tfplugindocs v0.8.1, so can you pull the latest changes from `future` and try again?

**Comment by alexjurkiewicz** [comment]
yup, LGTM!

**Review by marcinwyszynski — APPROVED** [review]

**Pull request merged**